Automattic/jetpack
# Slideshow block: Ensure max-width is defined when within the Stack block

Fixes https://github.com/Automattic/vulcan/issues/610

## Proposed changes:

* The issue: When a Slideshow block is added to a Stack block, it's max-width balloons, thanks to it's parent (the Stack block) having a display of flex set. This is a known issue with the Swiper library - see the link in the linked issue. This affects Slideshow blocks added in both templates and regular post / page editors, plus the block display on the front-end.
* This PR defines a max-width for the Slideshow block, when it is placed within a Stack block. 
* It also ensures the max-width considers any parent padding. Since this requires calculation and can't be done via pure CSS, a new `utils.js` file has been created with a JS helper function to help calculate and apply width changes based on padding. 
* I've tested putting the Slideshow block inside various other blocks, and the issue only occurs within the Stack block (the Row block with orientation set to vertical becomes a Stack block). It's possible that another parent-style block could also define itself as a flex container, however I've been unable to find an example. So until we can, I think the current approach would fit - if we find another flex parent block then we may need to change the logic so in JS we check for a flex parent and apply styles that way.
* A simpler solution would be great here - this seems overly complex, so I'm open to any other ideas as well.

A similar issue was fixed here, though was tested with Row / Column blocks: https://github.com/Automattic/jetpack/pull/40041


### Other information:


## Jetpack product discussion

https://github.com/Automattic/vulcan/issues/610
https://github.com/Automattic/jetpack/issues/22834
p1733937332404909-slack-C05PV073SG3

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

**To replicate the issue:**

On either self-hosted, WoA or Simple:

* Add a block-based theme (some that were mentioned in testing include Attar, Marl, Twenty Twenty Two, Blank Canvas).
* Open the site editor and open the 'index' template. Add a stack block, and then add the Slideshow block to that (with several images). You can move blocks around via the block overview section if needed.
* You should see part of the slideshow block but without any images - it's width extending well beyond the width of the screen.
* Open the site on the front-end to the homepage or the page that displays posts (the index page), and what you see should be similar.
* In the post / page editor for a post / page, add a Slideshow block within a Stack block - what you see should be similar again.
* Note there are other issues with the Slideshow block display (see below) - this largely depends on the theme and where you place the Slideshow, but for the purposes of this PR you may be better adding the Stack block within another such as the header, footer, or main content area.

**To test the fix**

On any platform:

* Apply this PR if on self-hosted locally, or via the Beta tester plugin on Jurassic Ninja or WoA, or using the generated commands in the comment below for a sandboxed Simple site.
* Follow the same steps as above.
* The slideshow block should be visible in those cases.
* Slideshows added to posts / pages outside of Stack blocks should continue to display as before.
* Test adding multiple slideshows to a post / page - for example a regular slideshow, a slideshow within a stack block, a slideshow within a stack block within a group block. Try adjusting image size as well (medium, large, etc). Everything should work as expected..

**Testing discrepancies:**
* Some template sections won't show the slideshow on the front-end, but they won't show other blocks either (not specific to slideshows).
* If you stack a stack block within a group block, you may also notice display issues. This should be fixable by modifying the group block / blocks layout options from the block sidebar. It may not be depending on how things are nested, but at this stage I'd think this would start becoming an edge case.
* You may also notice other unrelated Slideshow issues when adding blocks, such as this issue related to images appearing 'stuck' to the mouse: https://github.com/Automattic/jetpack/issues/37407

Screenshots:

Block overview section:
<img width="329" alt="Screenshot 2024-11-28 at 16 09 38" src="https://github.com/user-attachments/assets/4124881b-c255-41d2-87d1-bcfc02d6ebc8">

Template / post editor before:

<img width="794" alt="Screenshot 2024-11-28 at 16 13 42" src="https://github.com/user-attachments/assets/9fe05496-260d-4f22-b838-0cb71fcffeca">

Front-end before: 
<img width="1481" alt="Screenshot 2024-11-28 at 16 15 40" src="https://github.com/user-attachments/assets/55ff49bd-e1d1-4035-a8ab-02adb16f57d4">
